src/circuit/gadget/ecc/chip/mul/incomplete.rs

```diff
@@ -0,0 +1,418 @@
+
+            // q_mul == 1
+            let q_mul_one_checks = {
+                let q_mul_is_one =
+                    q_mul.clone() * (two.clone() - q_mul.clone()) * (three.clone() - q_mul.clone());
+                let y_a_witnessed = meta.query_advice(self.lambda1, Rotation::cur());
+                let y_a = y_a_next.clone();
+                Some(("init y_a", q_mul_is_one * (y_a - y_a_witnessed)))
+            };
+
+            // q_mul == 2
+            let q_mul_two_checks = {
+                let q_mul_is_two =
+                    q_mul.clone() * (one.clone() - q_mul.clone()) * (three - q_mul.clone());
+
+                // z_i
+                let z_cur = meta.query_advice(self.z, Rotation::cur());
+                // z_{i+1}
+                let z_prev = meta.query_advice(self.z, Rotation::prev());
+                // x_{A,i}
+                let x_a_cur = meta.query_advice(self.x_a, Rotation::cur());
+                // x_{A,i-1}
+                let x_a_next = meta.query_advice(self.x_a, Rotation::next());
+                // x_{P,i}
+                let x_p_cur = meta.query_advice(self.x_p, Rotation::cur());
+                // x_{P,i-1}
+                let x_p_next = meta.query_advice(self.x_p, Rotation::next());
+                // y_{P,i}
+                let y_p_cur = meta.query_advice(self.y_p, Rotation::cur());
+                // y_{P,i-1}
+                let y_p_next = meta.query_advice(self.y_p, Rotation::next());
+                // λ_{1,i}
+                let lambda1_cur = meta.query_advice(self.lambda1, Rotation::cur());
+                // λ_{2,i}
+                let lambda2_cur = meta.query_advice(self.lambda2, Rotation::cur());
+
+                // The current bit in the scalar decomposition, k_i = z_i - 2⋅z_{i+1}.
+                // Recall that we assigned the cumulative variable `z_i` in descending order,
+                // i from n down to 0. So z_{i+1} corresponds to the `z_prev` query.
+                let k = z_cur - z_prev * pallas::Base::from_u64(2);
+                // Check booleanity of decomposition.
+                let bool_check = k.clone() * (one.clone() - k.clone());
+
+                // The base used in double-and-add remains constant. We check that its
+                // x- and y- coordinates are the same throughout.
+                let x_p_check = x_p_cur.clone() - x_p_next;
+                let y_p_check = y_p_cur.clone() - y_p_next;
+
+                // λ_{1,i}⋅(x_{A,i} − x_{P,i}) − y_{A,i} + (2k_i - 1) y_{P,i} = 0
+                let poly1 = lambda1_cur * (x_a_cur.clone() - x_p_cur) - y_a_cur.clone()
+                    + (k * pallas::Base::from_u64(2) - one.clone()) * y_p_cur;
+
+                // λ_{2,i}^2 − x_{A,i-1} − x_{R,i} − x_{A,i} = 0
+                let secant_line = lambda2_cur.clone().square()
+                    - x_a_next.clone()
+                    - x_r(meta, Rotation::cur())
+                    - x_a_cur.clone();
```

Comment:
@ebfull is there any benefit to simplifying custom gates manually (eliminating `x_a_cur` here)? And/or can that simplification be done later if necessary, or automated, without changing `PinnedVerificationKey`?